File: stream/src/main/java/bsky4j/stream/internal/atproto/_SyncResource.java
package kbsky.stream.internal.atproto;

import kbsky.ATProtocol;
import kbsky.ATProtocolException;
import kbsky.ATProtocolTypes;
import kbsky.stream.api.atproto.SyncResource;
import kbsky.stream.api.entity.atproto.sync.SyncSubscribeReposRequest;
import kbsky.stream.util.StreamClient;

import java.net.URL;

public class _SyncResource implements SyncResource {

    private final ATProtocol atproto;
    private final String uri;

    _SyncResource(
            ATProtocol atproto,
            String uri
    ) {
        this.atproto = atproto;
        this.uri = uri;
    }

    @Override
    public StreamClient subscribeRepos(
            SyncSubscribeReposRequest request
    ) {
        try {
            String url = "wss://" + new URL(this.uri).getHost()
                    + "/xrpc/" + ATProtocolTypes.SyncSubscribeRepos;
            return new StreamClient(atproto, url);

        } catch (Exception e) {
            throw new ATProtocolException(e);
        }
    }
}

File: stream/src/commonMain/kotlin/work/socialhub/kbsky/stream/internal/atproto/_SyncResource.kt
package work.socialhub.kbsky.stream.internal.atproto

import io.ktor.http.*
import kbsky.stream.api.atproto.SyncResource
import work.socialhub.kbsky.ATProtocol
import work.socialhub.kbsky.ATProtocolTypes.SyncSubscribeRepos
import work.socialhub.kbsky.stream.api.entity.atproto.sync.SyncSubscribeReposRequest
import work.socialhub.kbsky.stream.util.StreamClient

class _SyncResource internal constructor(
    private val atproto: ATProtocol,
    private val uri: String
) : SyncResource {

    override fun subscribeRepos(
        request: SyncSubscribeReposRequest
    ): StreamClient {
        val url = ("wss://" + Url(this.uri).host
                + "/xrpc/" + SyncSubscribeRepos)
        return StreamClient(atproto, url, request.filter)
    }
}
